addons/binding/org.openhab.binding.freebox/src/main/java/org/openhab/binding/freebox/internal/FreeboxAirPlayAudioSink.java

```diff
@@ -0,0 +1,143 @@
+
+    private Set<AudioFormat> supportedFormats;
+    private AudioHTTPServer audioHTTPServer;
+    private FreeboxThingHandler handler;
+    private String callbackUrl;
+
+    public FreeboxAirPlayAudioSink(FreeboxThingHandler handler, AudioHTTPServer audioHTTPServer, String callbackUrl) {
+        this.handler = handler;
+        this.audioHTTPServer = audioHTTPServer;
+        this.callbackUrl = callbackUrl;
+        Boolean acceptLowBitrate = (Boolean) handler.getThing().getConfiguration()
+                .get(FreeboxAirPlayDeviceConfiguration.ACCEPT_ALL_MP3);
+        this.supportedFormats = new HashSet<>();
+        this.supportedFormats.add(WAV);
+        if (acceptLowBitrate) {
+            this.supportedFormats.add(MP3);
+        } else {
+            // Only accept MP3 bitrates >= 96 kbps
+            this.supportedFormats.add(MP3_96);
+            this.supportedFormats.add(MP3_112);
+            this.supportedFormats.add(MP3_128);
+            this.supportedFormats.add(MP3_160);
+            this.supportedFormats.add(MP3_192);
+            this.supportedFormats.add(MP3_224);
+            this.supportedFormats.add(MP3_256);
+            this.supportedFormats.add(MP3_320);
+        }
+        this.supportedFormats.add(OGG);
+    }
+
+    @Override
+    public String getId() {
+        return handler.getThing().getUID().toString();
+    }
+
+    @Override
+    public String getLabel(Locale locale) {
+        return handler.getThing().getLabel();
+    }
+
+    @Override
+    public void process(AudioStream audioStream) throws UnsupportedAudioFormatException {
+        if (!ThingHandlerHelper.isHandlerInitialized(handler)
+                || (handler.getThing().getStatus() != ThingStatus.ONLINE)) {
+            return;
+        }
+
+        String url = null;
+        if (audioStream instanceof URLAudioStream) {
+            // it is an external URL, we can access it directly
+            URLAudioStream urlAudioStream = (URLAudioStream) audioStream;
+            url = urlAudioStream.getURL();
+        } else {
+            if (callbackUrl != null) {
+                // we serve it on our own HTTP server
+                String relativeUrl;
+                if (audioStream instanceof FixedLengthAudioStream) {
+                    relativeUrl = audioHTTPServer.serve((FixedLengthAudioStream) audioStream, 20);
+                } else {
+                    relativeUrl = audioHTTPServer.serve(audioStream);
+                }
+                url = callbackUrl + relativeUrl;
+            } else {
+                logger.warn("We do not have any callback url, so AirPlay device cannot play the audio stream!");
+            }
+        }
+        try {
+            audioStream.close();
+        } catch (IOException e) {
+        }
+        try {
+            logger.debug("AirPlay audio sink: process url {}", url);
+            handler.playMedia(url);
+        } catch (Exception e) {
+            logger.warn("Audio stream playback failed", e);
```

Comment:
Is the exception stack trace relevant here? I.e. is this a bug situation where people need to provide the log for analysis?